File: apps/backend/src/scheduler.py
"""スケジュール管理モジュール"""

import os
import logging
from datetime import datetime, time, timedelta, timezone
from typing import Optional
from zoneinfo import ZoneInfo

logger = logging.getLogger(__name__)

JST = ZoneInfo("Asia/Tokyo")


class ScheduleConfig:
    """監視スケジュール設定"""

    def __init__(self):
        self.schedule_type = os.environ.get("SCHEDULE_TYPE", "always")
        self.schedule_days = self._parse_days(os.environ.get("SCHEDULE_DAYS", ""))
        self.start_time = self._parse_time(os.environ.get("SCHEDULE_START_TIME", "00:00"))
        self.duration_minutes = int(os.environ.get("SCHEDULE_DURATION_MINUTES", 1440))
        self.timezone = JST

        logger.info(
            f"Schedule config: type={self.schedule_type}, days={self.schedule_days}, "
            f"start={self.start_time}, duration={self.duration_minutes}min"
        )

    def _parse_days(self, days_str: str) -> list[int]:
        """曜日/日付文字列をパース"""
        if not days_str:
            return []

        day_map = {
            "mon": 0, "tue": 1, "wed": 2, "thu": 3,
            "fri": 4, "sat": 5, "sun": 6,
            "月": 0, "火": 1, "水": 2, "木": 3,
            "金": 4, "土": 5, "日": 6,
        }

        result = []
        for part in days_str.lower().split(","):
            part = part.strip()
            if part in day_map:
                result.append(day_map[part])
            elif part.isdigit():
                result.append(int(part))
        return result

    def _parse_time(self, time_str: str) -> time:
        """時刻文字列をパース (HH:MM形式)"""
        try:
            parts = time_str.split(":")
            return time(int(parts[0]), int(parts[1]))
        except (ValueError, IndexError):
            logger.warning(f"Invalid time format: {time_str}, using default")
            return time(0, 0)

    def is_active_now(self) -> bool:
        """現在時刻が監視対象期間かどうか"""
        return self.is_active_at(datetime.now(self.timezone))

    def _day_matches(self, date) -> bool:
        """指定した date がスケジュール対象の曜日/日付かどうか"""
        if not self.schedule_days:
            return True
        if self.schedule_type == "weekday":
            return date.weekday() in self.schedule_days
        if self.schedule_type == "day_of_month":
            return date.day in self.schedule_days
        return True

    def _find_schedule_start(self, dt: datetime) -> Optional[datetime]:
        """
        dt が含まれるスケジュール期間の開始日時を返す。
        当日・前日の start_time を候補として探す。
        """
        duration = timedelta(minutes=self.duration_minutes)
        for delta_days in (0, 1):
            candidate_date = (dt - timedelta(days=delta_days)).date()
            start = datetime.combine(candidate_date, self.start_time).replace(tzinfo=self.timezone)
            if start <= dt <= start + duration and self._day_matches(candidate_date):
                return start
        return None

    def is_active_at(self, dt: datetime) -> bool:
        """指定時刻が監視対象期間かどうか"""
        if self.schedule_type == "always":
            return True
        return self._find_schedule_start(dt) is not None

    def get_next_start(self) -> Optional[datetime]:
        """次回の収集開始時刻を返す (always モードは None)"""
        if self.schedule_type == "always":
            return None

        now = datetime.now(self.timezone)
        for delta in range(0, 14):
            candidate = (now + timedelta(days=delta)).replace(
                hour=self.start_time.hour,
                minute=self.start_time.minute,
                second=0, microsecond=0,
            )
            if candidate <= now:
                continue
            if self._day_matches(candidate):
                return candidate
        return None

    def get_status_message(self) -> str:
        """現在のスケジュール状態を説明するメッセージ"""
        if self.schedule_type == "always":
            return "Always monitoring"

        if self.schedule_type == "weekday":
            weekday_names = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
            days_str = ",".join(weekday_names[d] for d in self.schedule_days) if self.schedule_days else "all"
            msg = f"Weekdays: {days_str}"
        elif self.schedule_type == "day_of_month":
            days_str = ",".join(str(d) for d in self.schedule_days) if self.schedule_days else "all"
            msg = f"Days: {days_str}"
        else:
            msg = f"Custom: {self.schedule_type}"

        return f"{msg}, {self.start_time} + {self.duration_minutes}min JST"

File: apps/backend/src/scheduler.ts
const JST = "Asia/Tokyo";

type ScheduleType = "always" | "weekday" | "day_of_month";

const DAY_MAP: Record<string, number> = {
  mon: 0, tue: 1, wed: 2, thu: 3, fri: 4, sat: 5, sun: 6,
  月: 0, 火: 1, 水: 2, 木: 3, 金: 4, 土: 5, 日: 6,
};

function parseTime(str: string): { hour: number; minute: number } {
  const [h, m] = str.split(":").map(Number);
  return { hour: h ?? 0, minute: m ?? 0 };
}

function parseDays(str: string): number[] {
  if (!str) return [];
  return str.toLowerCase().split(",").flatMap((part) => {
    part = part.trim();
    if (part in DAY_MAP) return [DAY_MAP[part]!];
    const n = parseInt(part);
    return isNaN(n) ? [] : [n];
  });
}

function nowJst(): Date {
  // Returns a Date whose .getFullYear()/.getMonth() etc. are NOT in JST,
  // but we use Intl to extract JST components instead.
  return new Date();
}

function jstComponents(d: Date) {
  const fmt = new Intl.DateTimeFormat("en-US", {
    timeZone: JST,
    year: "numeric", month: "2-digit", day: "2-digit",
    hour: "2-digit", minute: "2-digit", second: "2-digit",
    hour12: false,
  });
  const parts = Object.fromEntries(fmt.formatToParts(d).map((p) => [p.type, p.value]));
  return {
    year: parseInt(parts.year!),
    month: parseInt(parts.month!) - 1,
    day: parseInt(parts.day!),
    hour: parseInt(parts.hour!),
    minute: parseInt(parts.minute!),
    second: parseInt(parts.second!),
    // weekday: 0=Mon ... 6=Sun (ISO)
    weekday: (d.getDay() + 6) % 7, // JS Sunday=0, convert to Mon=0
  };
}

function jstDateMs(year: number, month: number, day: number, hour: number, minute: number): number {
  // Create a Date representing the given wall-clock time in JST
  const utcStr = `${year}-${String(month + 1).padStart(2, "0")}-${String(day).padStart(2, "0")}T${String(hour).padStart(2, "0")}:${String(minute).padStart(2, "0")}:00+09:00`;
  return new Date(utcStr).getTime();
}

export class ScheduleConfig {
  readonly scheduleType: ScheduleType;
  readonly scheduleDays: number[];
  readonly startHour: number;
  readonly startMinute: number;
  readonly startTimeStr: string;
  readonly durationMinutes: number;

  constructor() {
    this.scheduleType = (process.env.SCHEDULE_TYPE ?? "always") as ScheduleType;
    this.scheduleDays = parseDays(process.env.SCHEDULE_DAYS ?? "");
    const t = parseTime(process.env.SCHEDULE_START_TIME ?? "00:00");
    this.startHour = t.hour;
    this.startMinute = t.minute;
    this.startTimeStr = `${String(t.hour).padStart(2, "0")}:${String(t.minute).padStart(2, "0")}`;
    this.durationMinutes = parseInt(process.env.SCHEDULE_DURATION_MINUTES ?? "1440");

    console.log(
      `[scheduler] type=${this.scheduleType} days=${this.scheduleDays} ` +
      `start=${this.startTimeStr} duration=${this.durationMinutes}min`
    );
  }

  private dayMatches(weekdayOrDay: number): boolean {
    if (this.scheduleDays.length === 0) return true;
    return this.scheduleDays.includes(weekdayOrDay);
  }

  private findScheduleStartMs(nowMs: number): number | null {
    const durationMs = this.durationMinutes * 60_000;
    for (const deltaDays of [0, 1]) {
      const candidateMs = nowMs - deltaDays * 86_400_000;
      const c = jstComponents(new Date(candidateMs));
      const startMs = jstDateMs(c.year, c.month, c.day, this.startHour, this.startMinute);
      const matchVal = this.scheduleType === "weekday" ? c.weekday : c.day;
      if (startMs <= nowMs && nowMs <= startMs + durationMs && this.dayMatches(matchVal)) {
        return startMs;
      }
    }
    return null;
  }

  isActiveNow(): boolean {
    if (this.scheduleType === "always") return true;
    return this.findScheduleStartMs(Date.now()) !== null;
  }

  getNextStart(): Date | null {
    if (this.scheduleType === "always") return null;

    const now = Date.now();
    for (let delta = 0; delta < 14; delta++) {
      const candidateMs = now + delta * 86_400_000;
      const c = jstComponents(new Date(candidateMs));
      const startMs = jstDateMs(c.year, c.month, c.day, this.startHour, this.startMinute);
      if (startMs <= now) continue;
      const matchVal = this.scheduleType === "weekday" ? c.weekday : c.day;
      if (this.dayMatches(matchVal)) return new Date(startMs);
    }
    return null;
  }

  statusMessage(): string {
    if (this.scheduleType === "always") return "Always monitoring";
    const weekdayNames = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"];
    const daysStr = this.scheduleDays.length > 0
      ? (this.scheduleType === "weekday"
        ? this.scheduleDays.map((d) => weekdayNames[d]).join(",")
        : this.scheduleDays.join(","))
      : "all";
    return `${this.scheduleType}: ${daysStr}, ${this.startTimeStr} +${this.durationMinutes}min JST`;
  }
}
